{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39d26e3d-da7e-403d-baf7-bbc9e68292f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import random\n",
    "import time\n",
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "from scipy.sparse import csr_matrix\n",
    "\n",
    "import pyranges as pr\n",
    "from pyranges.pyranges_main import PyRanges\n",
    "\n",
    "from utils import *\n",
    "from peak_parser import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51358891-1372-4717-9cb5-d772e46bd547",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ['CUDA_VISIBLE_DEVICES'] = '-1'  #ban GPU\n",
    "\n",
    "import tensorflow as tf\n",
    "import data_io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aed077d-f935-43ac-b2b4-31fef5cd97ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Custom TF peak length and epigenomic region\n",
    "#Extract DNA sequence at TF bingding sites (200bp)\n",
    "#\n",
    "cell='GM12878'\n",
    "tranf='CTCF'\n",
    "\n",
    "chroms=['chr1','chr2','chr3','chr4','chr5','chr6','chr7','chr8','chr9','chr10','chr11','chr12','chr13','chr14','chr15','chr16','chr17','chr18','chr19','chr20','chr21','chr22','chrX','chrY',]\n",
    "genome_fa_file='data/genome/GRCh38.primary_assembly.genome.fa' #downloaded from https://www.gencodegenes.org/\n",
    "\n",
    "path_peak='data/training'\n",
    "path_cell=path_peak+'/'+cell\n",
    "path_tf=path_cell+'/'+tranf\n",
    "tf_file=path_cell+'/'+tranf+'.bed'\n",
    "\n",
    "reg_file=path_tf+'/'+tranf+'_40k.bed'\n",
    "bed_file=path_tf+'/'+tranf+'_200.bed'\n",
    "reg_file_null=path_tf+'/'+tranf+'_null_40k.bed'\n",
    "bed_file_null=path_tf+'/'+tranf+'_null_200.bed'\n",
    "seq_file=path_tf+'/'+tranf+'_200.fa'\n",
    "seq_file_null=path_tf+'/'+tranf+'_null_200.fa'\n",
    "\n",
    "custom_bed(peak_file=tf_file,reg_file=reg_file,bed_file=bed_file,reg_file_null=reg_file_null,bed_file_null=bed_file_null,chroms=chroms)\n",
    "extract_fasta(bed_file=bed_file,fa_file=genome_fa_file,seq_file=seq_file)\n",
    "extract_fasta(bed_file=bed_file_null,fa_file=genome_fa_file,seq_file=seq_file_null)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8bcc7bf-d8e1-470e-991b-f99452fc99ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Negative FIMO sitesï¼Œgenerated by run_fimo.sh\n",
    "#\n",
    "fimo_neg_file=path_tf+'/'+'fimo_neg.bed'\n",
    "\n",
    "bedfile_fimo=path_tf+'/'+'fimo_neg_200.bed'\n",
    "regfile_fimo=path_tf+'/'+'fimo_neg_40k.bed'\n",
    "seqfile_fimo=path_tf+'/'+'fimo_neg_200.fa'\n",
    "\n",
    "fimo_neg=pr.read_bed(fimo_neg_file)\n",
    "fimo_neg = fimo_neg[fimo_neg.Chromosome.isin(chroms)]\n",
    "\n",
    "bed_region_fimo=custom_peak(fimo_neg,length=40000,write_file=regfile_fimo)\n",
    "custom_peak(bed_region_fimo,length=200,write_file=bedfile_fimo)\n",
    "extract_fasta(bed_file=bedfile_fimo,fa_file=genome_fa_file,seq_file=seqfile_fimo)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51292db3-691d-49ca-93ae-a3de0cf77c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert epigenomic information (EI) to state values (0,1) \n",
    "#\n",
    "epi_targets=['ATAC', 'H3K27ac', 'H3K27me3', 'H3K36me3', 'H3K4me1', 'H3K4me3', 'H3K9me3']  #\n",
    "\n",
    "peak_files={x:path_cell+'/'+x+'.bed' for x in epi_targets}\n",
    "ohpeak_files={x:path_cell+'/'+x+'.bed.pickle' for x in epi_targets}\n",
    "\n",
    "for k,v in peak_files.items():\n",
    "    if Path(v).exists():\n",
    "        print(v)\n",
    "        peak_2onehot_chrom_whole(v,ohpeak_files[k])\n",
    "    else:\n",
    "        print(v,'no such file.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5397336a-8f3d-4f6c-a8c7-5e814abfcd03",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Integrate DNA info. and EI info.\n",
    "#\n",
    "import shutil\n",
    "\n",
    "tmpdir=path_peak+'/'+'tmp'\n",
    "Path(tmpdir).mkdir(exist_ok=True)\n",
    "\n",
    "seq_file_pos=path_tf+'/'+tranf+'_200.fa'\n",
    "reg_file_pos=path_tf+'/'+tranf+'_40k.bed'\n",
    "sent_file_pos=path_tf+'/'+tranf+'_200.pickle'\n",
    "            \n",
    "seq_file_null=path_tf+'/'+tranf+'_null_200.fa'\n",
    "reg_file_null=path_tf+'/'+tranf+'_null_40k.bed'\n",
    "sent_file_null=path_tf+'/'+tranf+'_null_200.pickle'\n",
    "\n",
    "seq_file_neg=path_tf+'/'+'fimo_neg_200.fa'\n",
    "reg_file_neg=path_tf+'/'+'fimo_neg_40k.bed'\n",
    "sent_file_neg=path_tf+'/'+'fimo_neg_200.pickle'\n",
    "\n",
    "#Peak sites and null (random) sites\n",
    "generate_peak_context(seq_file=seq_file_pos,reg_file=reg_file_pos,label=1,targets=epi_targets,targets_files=ohpeak_files,out_file=sent_file_pos,tmpdir=tmpdir)\n",
    "generate_peak_context(seq_file=seq_file_null,reg_file=reg_file_null,label=0,targets=epi_targets,targets_files=ohpeak_files,out_file=sent_file_null,tmpdir=tmpdir)\n",
    "#Negative FIMO sites\n",
    "generate_peak_context(seq_file=seq_file_neg,reg_file=reg_file_neg,label=0,targets=epi_targets,targets_files=ohpeak_files,out_file=sent_file_neg,tmpdir=tmpdir)\n",
    "\n",
    "shutil.rmtree(tmpdir) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4cff6e3-d76d-4d6c-b924-4a2430a6d825",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Build tfrecord dataset, use all EI info.\n",
    "#\n",
    "sent_file_pos=path_tf+'/'+tranf+'_200.pickle'\n",
    "sent_file_null=path_tf+'/'+tranf+'_null_200.pickle'\n",
    "sent_file_neg=path_tf+'/'+'fimo_neg_200.pickle'\n",
    "\n",
    "file_peak_set=path_tf+'/'+'peak_set.tfrecord'\n",
    "file_train_set=path_tf+'/'+'train_set.tfrecord'\n",
    "file_test_set=path_tf+'/'+'test_set.tfrecord'\n",
    "    \n",
    "peak_set=data_io.select_sample(sample_file=sent_file_pos)  #no shuffling\n",
    "data_io.write_tfrecord(peak_set,file_peak_set) \n",
    "\n",
    "dataset=[]\n",
    "dataset+=data_io.select_sample(sample_file=sent_file_pos)\n",
    "dataset+=data_io.select_sample(sample_file=sent_file_null)\n",
    "dataset+=data_io.select_sample(sample_file=sent_file_neg)\n",
    "    \n",
    "random.shuffle(dataset)  #shuffle \n",
    "train_set,test_set=data_io.separate_dataset(dataset)  #\n",
    "    \n",
    "data_io.write_tfrecord(train_set,file_train_set)\n",
    "data_io.write_tfrecord(test_set,file_test_set)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ce46dbd-1446-42ca-baa8-f11fa554e8e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Build tfrecord dataset, use select EI info.\n",
    "epi_targets_comb={\n",
    "    'epi-only':'ATAC_H3K27ac_H3K27me3_H3K36me3_H3K4me1_H3K4me3_H3K9me3',\n",
    "    'dna-only':'',\n",
    "    'atac':'ATAC',\n",
    "    'h3k27ac':'H3K27ac',\n",
    "    'h3k27me3':'H3K27me3',\n",
    "    'h3k36me3':'H3K36me3',\n",
    "    'h3k4me1':'H3K4me1',\n",
    "    'h3k4me3':'H3K4me3',\n",
    "    'h3k9me3':'H3K9me3',\n",
    "    'bivalent':'H3K27me3_H3K4me3',\n",
    "    'epi-core':'H3K27me3_H3K36me3_H3K4me1_H3K4me3_H3K9me3'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5a95463-ca32-488b-909d-8ddfb1fb9aad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Taking ATAC as an example\n",
    "#\n",
    "target='atac'\n",
    "\n",
    "sent_file_pos=path_tf+'/'+tranf+'_200.pickle'\n",
    "sent_file_null=path_tf+'/'+tranf+'_null_200.pickle'\n",
    "sent_file_neg=path_tf+'/'+'fimo_neg_200.pickle'\n",
    "\n",
    "target_epis=epi_targets_comb[target].split('_')\n",
    "keep_dna=False if target=='epi-only' else True\n",
    "        \n",
    "file_peak_set=path_tf+'/'+'peak_set_'+target+'.tfrecord'\n",
    "file_train_set=path_tf+'/'+'train_set_'+target+'.tfrecord'\n",
    "file_test_set=path_tf+'/'+'test_set_'+target+'.tfrecord'\n",
    "\n",
    "peak_set=data_io.select_sample(sample_file=sent_file_pos)  #No shuffling\n",
    "peak_set=[mask_peak_context(x,target_epis,keep_dna=keep_dna) for x in peak_set]\n",
    "data_io.write_tfrecord(peak_set,file_peak_set)\n",
    "        \n",
    "dataset=[]\n",
    "dataset+=data_io.select_sample(sample_file=sent_file_pos)\n",
    "dataset+=data_io.select_sample(sample_file=sent_file_null)\n",
    "dataset+=data_io.select_sample(sample_file=sent_file_neg)\n",
    "random.shuffle(dataset)\n",
    "\n",
    "dataset=[mask_peak_context(x,target_epis,keep_dna=keep_dna) for x in dataset]\n",
    "train_set,test_set=data_io.separate_dataset(dataset)\n",
    "data_io.write_tfrecord(train_set,file_train_set)\n",
    "data_io.write_tfrecord(test_set,file_test_set)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23f49cb2-7205-4827-9726-987e02c50cd0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
